aspnet/mvc/models/validation.rst

```diff
@@ -1,8 +1,125 @@
+
+Below is an annotated ``Movie`` model from an app that stores information about movies and TV shows. Most of the fields are required, the ``Id`` field is a primary key, and several string properties have length requirements. Additionally, there is a numeric range in place for the ``Price`` property from 0 to $999.99, along with two custom annotations.
+
+.. literalinclude:: validation/sample/Movie.cs
+   :language: c#
+   :lines: 7-37
+   :dedent: 4
+
+Simply reading through the model reveals the rules about data for this app, making it easier to maintain the code. Below are several popular built-in data annotations:
+
+- ``[Bind]``: Determines whether the property is included or excluded from model binding.
+- ``[CreditCard]``: Formats the property as a credit card type.
+- ``[Compare]``: Compares two properties in a model. It works with various types, such as int, string, or date. 
+- ``[DataType]``: Applies extra type information to a property. For example, a string property may be formatted as an email data type.
+- ``[DisplayName]``: Determines the text that renders in form field labels and validation error messages.
+- ``[Email]``: Formats the property with an email format.
+- ``[ForeignKey]``: Provides foreign key functionality.
+- ``[Key]``: Provides primary key functionality. 
+- ``[Phone]``: Formats the property as a telephone format.
+- ``[Range]``: Sets the minimum and maximum values for multiple types of properties.
+- ``[RegularExpression]``: Forces the data to match a regular expression.
+- ``[Required]``: Makes a property required.
+- ``[ScaffoldColumn]``: Determines whether to render the property it is applied to.
+- ``[StringLength]``: The maximum length for a string field.
+- ``[Url]``: Formats the property as a URL.
+
+A complete list of annotations is at the `System.ComponentModel.DataAnnotations <https://msdn.microsoft.com/en-us/library/system.componentmodel.dataannotations(v=vs.110).aspx>`_ namespace docs. Note that this namespace is not specific to MVC, but in a shared .NET namespace. 
+
+Many annotations provide actions similar to constraint checking in database tables, or in popular business logic patterns. However, there are intances where business rules don't fit into these common scenarios. For those times, you can create custom validation attributes to apply to model properties.  
+
+Model State
+-----------
+Model state gives you insight as to which form fields have not passed validation. Controllers contain a ``ModelState`` property that is of type ``ModelStateEntry`` and it contains an ``Errors`` property. The errors are a list of keys and values that contains the error messages and other information. You can use the error information to build a response so that the client will know which fields did not pass validation. In the case of a validation error, APIs should return an HTTP status code ``400 Bad Request``.
```

Comment:
You can just point to the API ref docs for the [ModelState](https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Controller/index.html#prop-Microsoft.AspNet.Mvc.Controller.ModelState) instead of trying to describe the type structure.